### PR TITLE
Smarter balance fetching

### LINKED ISSUE #3240: UI causes Parity to consume too much CPU
Both Parity & Chrome shows higher than normal CPU usage when the app is open. Check -

- status optimisations (slower, not all)
- running timers, polling

### LINKED ISSUE #3590: Make balance fetching context specific
Accounts balances fetch should be smarter.

Don't query transaction count if not on an account page (won't be used anyway).
Query balances only for visible accounts.
Try to use `Transfer` event to query tokens' balance on new block (refresh every balance once in a while though).

### PR DESCRIPTION
**Pull request by ngotchac**
Closes #3590 

This might also (hopefully) close #3240 

Fetching balances is done smarter, in several ways:

- Only fetch balances of currently visible accounts
- Fetch token balances with only 2 calls via eth Filters (2 because of FROM and TO) and the Transfer event

It is likely that every token used will implement the correct Transfer event. Some implementations might differ, or for example when buying some GavCoin (tokens created). Thus all the balances are fetched every 2 minutes, just to be sure.

### CODE REVIEW

File: js/src/api/format/input.js [hunk 1]
```diff
@@ -51,15 +51,30 @@ export function inHash (hash) {
   return inHex(hash);
 }
 
-export function pad (input, length) {
+export function padRight (input, length) {
```
Comment (jacogr) [inline js/src/api/format/input.js]: Move to `/api/util/format.js`? (pad{left,Right})

File: js/src/redux/providers/balances.js [hunk 2]
```diff
@@ -16,68 +16,63 @@
...
+    this.shortThrottledFetch = throttle(
+      this.fetchBalances,
+      2 * 1000,
+      { trailing: true }
+    );
+
+    // Fetch all tokens every 2 minutes
+    this.throttledTokensFetch = throttle(
+      this.fetchTokens,
+      2 * 60 * 1000,
```
Comment (jacogr) [inline js/src/redux/providers/balances.js]: Start at 60s

File: js/src/redux/providers/balancesActions.js [hunk 3]
```diff
@@ -14,16 +14,355 @@
...
+    ])
+    .then(([ tokenData, image ]) => {
+      const [ address, tag, format, name ] = tokenData;
+      const contract = api.newContract(ABIS.eip20, address);
+
+      const token = {
+        format: format.toString(),
+        id: tokenId,
+        image: hashToImageUrl(image),
+
```
Comment (jacogr) [inline js/src/redux/providers/balancesActions.js]: extra line to be removed?

File: js/src/redux/providers/balancesReducer.js [hunk 4]
```diff
@@ -15,22 +15,95 @@
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
 import { handleActions } from 'redux-actions';
+import BigNumber from 'bignumber.js';
 
 const initialState = {
   balances: {},
-  tokens: {}
+  tokens: {},
+
```
Comment (jacogr) [inline js/src/redux/providers/balancesReducer.js]: extra line to be removed?

File: js/src/redux/providers/balancesReducer.js [hunk 5]
```diff
@@ -15,22 +15,95 @@
...
+    const balances = { ...prevBalances };
+
+    Object.keys(nextBalances).forEach((address) => {
+      if (!balances[address]) {
+        balances[address] = Object.assign({}, nextBalances[address]);
+        return;
+      }
+
+      const balance = Object.assign({}, balances[address]);
+
```
Comment (jacogr) [inline js/src/redux/providers/balancesReducer.js]: extra blank (and 2 down as well, doesn't add much in terms of readability)

File: js/src/views/Dapps/dappsStore.js [hunk 6]
```diff
@@ -249,12 +251,27 @@ export default class DappsStore {
   }
 
   _fetchManifest (manifestHash) {
+    if (/^(0x)?0+/.test(manifestHash)) {
+      return Promise.resolve(null);
+    }
+
+    if (this._manifests[manifestHash]) {
+      return this._manifests[manifestHash];
```
Comment (jacogr) [inline js/src/views/Dapps/dappsStore.js]: Promise.resolve?

**Comment by ngotchac** [comment]
@jacogr Should be ok now :)

**Pull request merged**